### PR TITLE
Expand security considerations, split into sub-sections

### PR DESCRIPTION
**Pull request by kvark**
This material was prepared in collaboration with @jdashg (thank you!).
It expands the security section and renames it to encompass the other aspects of malicious use. We may consider moving it to an appendix.

***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/898.html" title="Last updated on Jul 6, 2020, 8:39 PM UTC (3627da4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/898/c66dabc...kvark:3627da4.html" title="Last updated on Jul 6, 2020, 8:39 PM UTC (3627da4)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -155,22 +159,40 @@ See [[#errors-and-debugging]] for more information about error handling.
...
 In order to address that, the shader instruction set and its defined behaviors are
-strictly defined by WebGPU. When a shader is provided, the WebGPU implementation has to validate it
+strictly defined by WebGPU. When a shader is provided to {{GPUDevice/createShaderModule()}},
+the WebGPU implementation has to validate it
 before doing any translation (to platform-specific shaders) or transformation passes.
 
+## Uninitalized data ## {#security-uninitialized}
+
+Generally, allocating new memory may expose the leftover data of other applications running on the system.
+In order to address that, WebGPU initializes all the resources to zero.
```
Comment (Kangz) [inline spec/index.bs:169]: Should we mention something about the clearing being expected to be done lazily?

File: spec/index.bs [hunk 2]
```diff
@@ -212,6 +234,18 @@ from multiple Web Workers, which makes the thread scheduling observable
 and allows the creation of high-precision timers.
 The theoretical attack vectors are a subset of those of SharedArrayBuffer.
 
+WebGPU also specifies the {{GPUExtensionName/"timestamp-query"}} extension, which
+provides high precision timing of GPU operations. The extension is optional, and a WebGPU
+implementation may limit its exposure only to those scenarios that trusted. Alternatively,
```
Comment (Kangz) [inline spec/index.bs:239]: nit: that are trusted

File: spec/index.bs [hunk 3]
```diff
@@ -220,11 +254,76 @@ For GPU processing time, a WebGPU implementation may set up "watchdog" timer tha
...
+mitigations for this.
+
+### Memory resources ### {#security-memory-resources}
+
+WebGPU exposes fallible allocations from machine-global memory heaps, such as VRAM.
+This allows for probing the size of the system's remaining available memory
+(for a given heap type) by attempting to allocate and watching for allocation failures.
+
+GPUs internally have one or more (typically only two) heaps of memory
+hared by all running applications. When a heap is depleted, WebGPU would fail to create a resource.
```
Comment (Kangz) [inline spec/index.bs:274]: nit: shared

File: spec/index.bs (hunk 1 above)
Comment (kvark) [reply spec/index.bs]: Should be better now. Please re-open if more changes are needed.

File: spec/index.bs [hunk 4]
```diff
@@ -155,22 +159,41 @@ See [[#errors-and-debugging]] for more information about error handling.
...
 
+## Uninitalized data ## {#security-uninitialized}
+
+Generally, allocating new memory may expose the leftover data of other applications running on the system.
+In order to address that, WebGPU conceptually initializes all the resources to zero, although in practice
+an implementation may skip this step if it sees the developer initializing the contents manually.
+This includes variables and shared threadgroup memory inside shaders.
+
+The precise mechanism of clearing the threadgroup memory can differ between platforms.
+If the native API does not provide facilities to clear it, WebGPU implementation transforms the compute
```
Comment (JusSn) [inline spec/index.bs:174]:
```suggestion
If the native API does not provide facilities to clear it, the WebGPU implementation transforms the compute
```

File: spec/index.bs [hunk 5]
```diff
@@ -220,11 +255,76 @@ For GPU processing time, a WebGPU implementation may set up "watchdog" timer tha
...
+are used by other applications, and how much they allocate from them.
+
+### Computation resources ### {#security-computation-resources}
+
+If one site uses WebGPU at the same time as another, it may observe the increase
+in time it takes to process some work. For example, if a site constantly submits
+compute workloads and tracks fences for their completion,
+it may observe that something else also started using the GPU.
+
+GPU has many parts that can be tested independently, such as the arithmetic units,
```
Comment (JusSn) [inline spec/index.bs:286]:
```suggestion
A GPU has many parts that can be tested independently, such as the arithmetic units,
```

File: spec/index.bs [hunk 6]
```diff
@@ -220,11 +255,76 @@ For GPU processing time, a WebGPU implementation may set up "watchdog" timer tha
...
+
+WebGPU is designed to have a strong baseline for the limits. This makes it viable (unlike WebGL)
+to *only* expose the baseline limits, trivially reducing the fingerprinting surface.
+The implementations may decide to only expose the baseline even if the hardware is more capable,
+or to support a limited number of higher-capable feature sets.
+
+### Machine-specific artifacts ### {#security-machine-artifacts}
+
+WebGPU doesn't specify some of the hardware behavior down to the expected bit values,
+providing some leeway for the platforms to differenciate. This applies to rasterization coverage
```
Comment (JusSn) [inline spec/index.bs:309]:
```suggestion
providing some leeway for the platforms to differentiate. This applies to rasterization coverage
```

**Comment by kvark** [comment]
Agreed to land and iterate in follow-ups

**Pull request merged**